arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stream/callbacks.kt

```diff
@@ -0,0 +1,124 @@
+package arrow.fx.coroutines.stream
+
+import arrow.core.Either
+import arrow.fx.coroutines.CancelToken
+import arrow.fx.coroutines.ExitCase
+import arrow.fx.coroutines.UnsafePromise
+import arrow.fx.coroutines.andThen
+import arrow.fx.coroutines.stream.concurrent.Queue
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.startCoroutine
+import kotlin.experimental.ExperimentalTypeInference
+
+private object END
+
+interface EmitterSyntax<A> {
+  fun emit(a: A): Unit
+  fun emit(chunk: Chunk<A>): Unit
+  fun emit(iterable: Iterable<A>): Unit
+  fun emit(vararg aas: A): Unit
+  fun end(): Unit
+}
+
+/**
+ * Creates a Stream from the given suspended block.
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.coroutines.stream.*
+ *
+ * //sampleStart
+ * suspend fun main(): Unit =
+ *   Stream.async {
+ *       emit(1)
+ *       emit(2, 3, 4)
+ *       end()
+ *     }
+ *     .compile()
+ *     .toList()
+ *     .let(::println) //[1, 2, 3, 4]
+ * //sampleEnd
+ * ```
+ *
+ * Note that if neither `end()`, `emit(Chunk.empty())` nor other limit operators such as `take(N)` are called,
+ * then the Stream will never end.
+ */
+// @OptIn(ExperimentalTypeInference::class) in 1.3.70
+@UseExperimental(ExperimentalTypeInference::class)
+fun <A> Stream.Companion.async(@BuilderInference f: suspend EmitterSyntax<A>.() -> Unit): Stream<A> =
+  Stream.asyncCancellable(f.andThen { CancelToken.unit })
+
+/**
+ * Creates a Stream from the given suspended block that will evaluate the passed CancelToken if cancelled.
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.coroutines.stream.*
+ * import arrow.fx.coroutines.CancelToken
+ *
+ * //sampleStart
+ * suspend fun main(): Unit =
+ *   Stream.asyncCancellable {
+ *       emit(1)
+ *       emit(2, 3, 4)
+ *       end()
+ *       CancelToken { /* cancel subscription to callback */ }
+ *     }
+ *     .compile()
+ *     .toList()
+ *     .let(::println) //[1, 2, 3, 4]
+ * //sampleEnd
+ * ```
+ *
+ * Note that if neither `end()`, `emit(Chunk.empty())` nor other limit operators such as `take(N)` are called,
+ * then the Stream will never end.
+ */
+// @OptIn(ExperimentalTypeInference::class) in 1.3.70
+@UseExperimental(ExperimentalTypeInference::class)
+fun <A> Stream.Companion.asyncCancellable(@BuilderInference f: suspend EmitterSyntax<A>.() -> CancelToken): Stream<A> =
```

Comment:
```suggestion
fun <A> Stream.Companion.cancellable(@BuilderInference f: suspend EmitterSyntax<A>.() -> CancelToken): Stream<A> =
```

We should rename this to be `cancellable` such as that it follows the same familira `Async/Concurrent` naming.